ui-speedspacechart/src/components/helpers/drawElements/powerRestrictions.ts

```diff
@@ -0,0 +1,138 @@
+    ratioX,
+    leftOffset,
+    layersDisplay: { electricalProfiles },
+  } = store;
+
+  clearCanvas(ctx, width, LINEAR_LAYERS_HEIGHTS.POWER_RESTRICTIONS_HEIGHT);
+
+  ctx.save();
+  ctx.translate(leftOffset, 0);
+
+  const { maxPosition } = maxPositionValues(store);
+
+  ctx.font = 'normal 12px IBM Plex Sans';
+  ctx.fillStyle = 'rgb(121, 118, 113)';
+  ctx.strokeStyle = 'rgb(121, 118, 113)';
+  ctx.lineWidth = 1;
+  ctx.textAlign = 'center';
+
+  let powerRestrictionsBackgroundColor = LINEAR_LAYERS_BACKGROUND_COLOR.FIRST;
+
+  // Power restrictions linear can either be the first layer
+  // if electrical profiles aren't displayed or second if they are
+  if (electricalProfiles) {
+    powerRestrictionsBackgroundColor = LINEAR_LAYERS_BACKGROUND_COLOR.SECOND;
+  }
+
+  drawSeparatorLinearLayer(
+    ctx,
+    'rgba(0,0,0,0.1)',
+    MARGINS,
+    width,
+    marginTop - LINEAR_LAYERS_HEIGHTS.POWER_RESTRICTIONS_HEIGHT + LINEAR_LAYER_SEPARATOR_HEIGHT
+  );
+
+  drawLinearLayerBackground(
+    ctx,
+    powerRestrictionsBackgroundColor,
+    MARGINS,
+    width,
+    marginTop,
+    LINEAR_LAYERS_HEIGHTS.POWER_RESTRICTIONS_HEIGHT - LINEAR_LAYER_SEPARATOR_HEIGHT
+  );
+
+  powerRestrictions.forEach((range, i) => {
+    if (range.handled) {
+      ctx.strokeStyle = 'rgb(121, 118, 113)';
+      ctx.fillStyle = 'rgb(121, 118, 113)';
+    } else {
+      ctx.strokeStyle = 'rgb(217, 28, 28)';
+      ctx.fillStyle = 'rgb(217, 28, 28)';
+    }
+    // Draw vertical line
+    const startX = positionOnGraphScale(range.start, maxPosition, width, ratioX, MARGINS);
+    const verticalStartY = LEFT_VERTICAL_LINE_PADDING_TOP;
+    ctx.beginPath();
+    ctx.moveTo(startX, verticalStartY);
+    ctx.lineTo(startX, verticalStartY + LEFT_VERTICAL_LINE_HEIGHT);
+
+    // Draw horizontal line around text
+    const textWidth = Math.floor(ctx.measureText(range.code).width);
+    let horizontalEndX = positionOnGraphScale(range.stop, maxPosition, width, ratioX, MARGINS);
+    if (i === powerRestrictions.length - 1) {
+      horizontalEndX -= MARGINS.CURVE_MARGIN_SIDES / 2;
+    }
+    const horizontalY = verticalStartY + HORIZONTAL_LINE_PADDING_TOP;
+
+    if (horizontalEndX - startX < textWidth + TEXT_PADDING_HORIZONTAL * 2 + TEXT_DISPLAY_MARGIN) {
+      // Hide text if line is too short
+      ctx.moveTo(startX, horizontalY);
+      ctx.lineTo(horizontalEndX - 2, horizontalY);
+    } else {
+      const textStartX = (startX + horizontalEndX) / 2 - textWidth / 2 - TEXT_PADDING_HORIZONTAL;
+      const textEndX = (startX + horizontalEndX) / 2 + textWidth / 2 + TEXT_PADDING_HORIZONTAL;
+
+      ctx.moveTo(startX, horizontalY);
+      ctx.lineTo(textStartX - TEXT_PADDING_HORIZONTAL, horizontalY);
+      ctx.moveTo(textEndX + TEXT_PADDING_HORIZONTAL, horizontalY);
+      ctx.lineTo(horizontalEndX - 2, horizontalY);
+
+      const textX = (startX + horizontalEndX) / 2;
+      ctx.fillText(range.code, textX, horizontalY + 4); // text centering adjustment
```

Comment:
Would we be able to leverage [`textBaseline`](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/textBaseline) to drop this constant?

Reply:
Done

Reply:
It seems like [`textBaseline`](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/textBaseline) isn't used in the last push? Did this end up not working?

Reply:
To be clear, my suggestion here is to add something like `ctx.textBaseline = 'middle'`.

Reply:
Oh got it, I'll check that

Reply:
The `textBaseline` solution might look better when some of the characters' shapes are under the baseline, such as "p" or "g". The "F" and "P" characters' shapes are all above the baseline (they take up a lot more space on the top).

Reply:
Hm, actually it seems like it's the other way around. Or maybe adding `horizontalY + 0.5` would help? (half the `lineWidth`)

Reply:
Better yet, `horizontalY + ctx.lineWidth / 2` to better reflect the intent

Reply:
Oh yep, it seems to be perfect like that !

Reply:
Done